jetty-http/src/main/java/org/eclipse/jetty/http/GZIPContentDecoder.java

```diff
@@ -0,0 +1,415 @@
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.http;
+
+import java.nio.ByteBuffer;
+import java.util.zip.DataFormatException;
+import java.util.zip.Inflater;
+import java.util.zip.ZipException;
+
+import org.eclipse.jetty.io.ByteBufferPool;
+import org.eclipse.jetty.util.BufferUtil;
+
+/**
+ * Decoder for the "gzip" encoding.
+ * <p>An Decode that inflates gzip compressed data that has been
+ * optimized for async usage with minimal data copies.
+ * 
+ */
+public class GZIPContentDecoder
+{
+    private final Inflater _inflater = new Inflater(true);
+    private final ByteBufferPool _pool;
+    private final int _bufferSize;
+    private State _state;
+    private int _size;
+    private int _value;
+    private byte _flags;
+    private ByteBuffer _empty;
+    private ByteBuffer _inflated;
+
+    public GZIPContentDecoder()
+    {
+        this(null,2048);
+    }
+
+    public GZIPContentDecoder(int bufferSize)
+    {
+        this(null,bufferSize);
+    }
+    
+    public GZIPContentDecoder(ByteBufferPool pool, int bufferSize)
+    {
+        _bufferSize = bufferSize;
+        _pool = pool;
+        reset();
+    }
+
+    /** Inflate compressed data from a buffer.
+     * 
+     * @param compressed Buffer containing compressed data.
+     * @return Buffer containing inflated data.
+     */
+    public ByteBuffer decode(ByteBuffer compressed)
+    {
+        decodeChunks(compressed);
+        if (BufferUtil.isEmpty(_inflated) || _state==State.CRC || _state==State.ISIZE )
+            return BufferUtil.EMPTY_BUFFER;
+        
+        ByteBuffer result = _inflated;
+        _inflated = null;
+        return result;
+    }
+
+    /** Called when a chunk of data is inflated.
+     * <p>The default implementation aggregates all the chunks 
+     * into a single buffer returned from {@link #decode(ByteBuffer)}.
+     * Derived implementations may choose to consume chunks individually
+     * and return false to prevent further inflation until a subsequent
+     * call to {@link #decode(ByteBuffer)} or {@link #decodeChunks(ByteBuffer)}.
+     *  
+     * @param chunk The inflated chunk of data
+     * @return False if inflating should continue, or True if the call 
+     * to {@link #decodeChunks(ByteBuffer)} or {@link #decode(ByteBuffer)} 
+     * should return, allowing back pressure of compressed data.
+     */
+    protected boolean decodedChunk(ByteBuffer chunk)
```

Comment:
This method allows back pressure during inflation by returning true.  Previous impl always fully inflated.... perhaps the client could also use this?